compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp

```diff
@@ -91,6 +92,40 @@ static bool isTransposeLinalgOp(linalg::LinalgOp linalgOp) {
   return isInputTransposed || isOutputTransposed;
 }
 
+/// Returns true if the index map represents a transpose that benefits from
+/// shared mem.
+static bool isSharedMemTranspose(AffineMap indexMap) {
+  if (!indexMap.isEmpty() && indexMap.isPermutation()) {
+    // Ensure that the fasted moving dimension (the last one) is permuted,
+    // Otherwise shared memory promotion will not benefit the operation.
+    if (indexMap.getDimPosition(indexMap.getNumDims() - 1) !=
+        indexMap.getNumDims() - 1) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Checks preconditions for shared mem transpose.
+static bool checkTransposePreconditions(LinalgOp linalgOp) {
+  // Check that the op has at least 2 to 3 parallel loops.
+  if (linalgOp.getNumParallelLoops() < 2 ||
+      linalgOp.getNumParallelLoops() > 3) {
+    return false;
+  }
+
+  // Check that all the iterators are parallel.
+  if (linalgOp.getNumParallelLoops() != linalgOp.getNumLoops()) {
+    return false;
+  }
+
+  // Only transpose static shapes
+  if (linalgOp.hasDynamicShape()) {
+    return false;
+  }
+  return true;
+}
```

Comment:
These also look like specific constraints for your case as a transpose can have more than 3 dimensions, dynamic shapes, reduction dimensions, etc. We can add another `[](LinalgOp op){}` lambda, to filter out LinalgOps, in this case.